worlds/mm2/Rules.py

```diff
@@ -0,0 +1,295 @@
+                if boss_damage[0]:
+                    boss_health[boss] = 0  # if we can buster, we should buster
+                    continue
+                highest, wp = max(zip(weapon_weight.values(), weapon_weight.keys()))
+                uses = weapon_energy[wp] // weapon_costs[wp]
+                if int(uses * boss_damage[wp]) > boss_health[boss]:
+                    used = ceil(boss_health[boss] / boss_damage[wp])
+                    weapon_energy[wp] -= weapon_costs[wp] * used
+                    boss_health[boss] = 0
+                elif highest <= 0:
+                    # we are out of weapons that can actually damage the boss
+                    # so find the weapon that has the most uses, and apply that as an additional weakness
+                    # it should be impossible to be out of energy, simply because even if every boss took 1 from
+                    # Quick Boomerang and no other, it would only be 28 off from defeating all 9, which Metal Blade should
+                    # be able to cover
+                    wp, max_uses = max((weapon, weapon_energy[weapon] // weapon_costs[weapon]) for weapon in weapon_weight
+                                       if weapon != 0)
+                    world.weapon_damage[wp][boss] = minimum_weakness_requirement[wp]
+                    used = min(int(weapon_energy[wp] // weapon_costs[wp]),
+                               ceil(boss_health[boss] // minimum_weakness_requirement[wp]))
+                    weapon_energy[wp] -= weapon_costs[wp] * used
+                    boss_health[boss] -= int(used * minimum_weakness_requirement[wp])
+                    weapon_weight.pop(wp)
+                else:
+                    # drain the weapon and continue
+                    boss_health[boss] -= int(uses * boss_damage[wp])
+                    weapon_energy[wp] -= weapon_costs[wp] * uses
+                    weapon_weight.pop(wp)
+
+    for i, boss_locations in zip(range(14), [
+        heat_man_locations,
+        air_man_locations,
+        wood_man_locations,
+        bubble_man_locations,
+        quick_man_locations,
+        flash_man_locations,
+        metal_man_locations,
+        crash_man_locations,
+        wily_1_locations,
+        wily_2_locations,
+        wily_3_locations,
+        wily_4_locations,
+        wily_5_locations,
+        wily_6_locations
+    ]):
+        if world.weapon_damage[0][i] > 0:
+            continue  # this can always be in logic
+        weapons = []
+        for weapon in range(1, 9):
+            if world.weapon_damage[weapon][i] > 0:
+                if world.weapon_damage[weapon][i] < minimum_weakness_requirement[weapon]:
+                    continue  # Atomic Fire can only be considered logical for bosses it can kill in 2 hits
+                weapons.append(weapons_to_name[weapon])
+        if not weapons:
+            raise Exception(f"Attempted to have boss {i} with no weakness! Seed: {world.multiworld.seed}")
```

Comment:
When genning with a yaml with `random_weakness` set to `shuffled` in 100-player multiworlds, it frequently fails here, to the point that I have not had a successful generation like that.

Failing yaml attached (converted to txt so I can post it here)
[weights.txt](https://github.com/user-attachments/files/16352643/weights.txt)